dags/module/coin_price_api_call.py

```diff
@@ -482,134 +449,367 @@ def process_data(
-            # timedelta(hours=1)로 하니까 정각에 호출하게 돼서 그 시간의 데이터를 가져오지못함. -> 1분 뒤의 시간으로 호출
-            start_date = last_loaded_time + timedelta(hours=1, minutes=1)
-            to_date = start_date.strftime("%Y-%m-%dT%H:%M:%S")
-            print(f"to date: {to_date}")
-            data = fetch_ohlcv_data("KRW-BTC", to_date, 1)
-            initial_load = False
-
-        # 그 외에는 최초삽입에 해당하므로 1년치 데이터 호출
-        else:
-            print("load first. fetch 365days data")
-            start_date = end_date - timedelta(days=365)
-            existing_times = set(time[0] for time in session.query(BtcOhlc.time).all())
-            data = fetch_data(start_date, end_date, existing_times)
-            # 중복 제거(check_and_remove_duplicates 함수는 sqlalchemy orm으로 작성되어서 테이블에 데이터가 존재하지 않을 시 중복제거 작업이 불가능)
-            # 해쉬테이블을 이용해 제거
-            unique_data = {entry["candle_date_time_kst"]: entry for entry in data}
-            data = list(unique_data.values())
-            initial_load = True
-
-        # 세션 내 데이터 확인
-        session_times = session.query(BtcOhlc.time).all()
-        print(f"Data in session before processing: {session_times}")
-        s_time = time.time()
-        print(f"1234134: {data[0]}")
-        if data:
-            process_data(data, start_date, end_date, session, initial_load)
-            print("data processing finished")
-        else:
-            print("No data fetched")
-        check_and_sort_data(session)
-        e_time = time.time()
-        etime = e_time - s_time
-        print(f"processing time: {etime} seconds")
+        threshold_date = datetime.utcnow() + timedelta(hours=9) - timedelta(days=365)
+        deleted_rows = (
+            session.query(BtcOhlcv).filter(BtcOhlcv.time < threshold_date).delete()
+        )
+        session.commit()
+        logger.info(f"Deleted {deleted_rows} old records from the database.")
     except Exception as e:
-        print(f"An error occurred: {e}")
         session.rollback()
+        logger.error(f"Failed to delete old data: {e}")
+    finally:
+        session.close()
+
+
+async def insert_data_into_db(fetch_data):
+    """
+    가져온 데이터를 데이터베이스에 삽입하는 함수
+
+    Args:
+        fetch_data (list): 가져온 데이터 목록
+    """
+    postgres_hook = PostgresHook(postgres_conn_id=Connections.POSTGRES_DEFAULT.value)
+    engine = create_engine(postgres_hook.get_uri())
+    SessionLocal = sessionmaker(bind=engine)
+
+    def is_continuous_timestamps(data):
+        """
+        데이터의 타임스탬프가 1시간 간격으로 연속적인지 확인하는 함수
 
+        Args:
+            data (list): 데이터 목록
+
+        Returns:
+            tuple: (연속적인지 여부, 마지막 연속된 타임스탬프, 연속되지 않은 첫 타임스탬프)
+        """
+        timestamps = [
+            datetime.strptime(entry["candle_date_time_kst"], "%Y-%m-%dT%H:%M:%S")
+            for entry in data
+        ]
+        for i in range(1, len(timestamps)):
+            if (timestamps[i] - timestamps[i - 1]) != timedelta(hours=1):
+                return False, timestamps[i - 1], timestamps[i]
```

Comment:
현재 python 반복문을 활용하여 시간의 연속성을 검사하고 계신데 이렇게 하면 작은 데이터에는 효율적일 수 있지만, 우리 프로젝트에서는 8000개 이상의 데이터를 다루기 때문에 인덱스를 사용한 쿼리 및 pandas, numpy을 사용해서 확인하는게 더  효율적일 것으로 예상됩니다. 코드 제안드립니다.
```suggestion
                time_series = pd.DataFrame(timestamps, columns=['time'])               
                # 1시간간격 타임스태프 생성
                times = pd.date_range(start=start_date, end=end_date, freq='h')
                # 누락된 시간 찾기(numpy사용)
                times = timestamps.to_numpy()
                actual_times = time_series['time'].sort_values().to_numpy()
                missing_times = np.setdiff1d(times, actual_times)
                #검사
                if len(missing_times) == 0:
                      return True, None, None
                else:
                      return False, None, None
```
start_data, end_date 는 따로 불러와서 넣어주시면 될 것 같습니다.

Reply:
마지막 코드 제안부분 557~559번줄 까지 입니다 범위를 잘못 체크했어요
@tfedohk

Reply:
@p65454 
```python
pd.DataFrame(timestamps, columns=['time’]):
```
이 부분의 시간 복잡도는 O(n:타임스탬프)입니다.

```python
pd.date_range(start=start_date, end=end_date, freq=‘H’):
```
이 부분은 주어진 시작 날짜와 종료 날짜 사이에 1시간 간격으로 타임스탬프를 생성하는 부분인데, 우리 데이터는 전체 기간에 대한 타임스탬프를 생성하는 것이므로, 결론적으로 앞선 시간 복잡도와 동일한 O(n:타임스탬프)입니다.

```python
time_series['time'].sort_values().to_numpy():
```
정렬 알고리즘은 일반적으로 O(n log n) 시간 복잡도를 보입니다.

```python
np.setdiff1d(times.to_numpy(), actual_times):
```
times와 actual_times는 결론적으로는 모두 n개의 타임스탬프입니다. 
시간 복잡도는 O(n log n)입니다.

최종적인 시간복잡도는 O(n) + O(n) + O(n log n)으로서, 곧 O(n log n)입니다.
따라서 기존 코드가 효율성 측면에서 더 나은 것으로 보입니다.
다시 한 번 검토해주세요.

Reply:
음 다시 생각해보니  굳이 바꿀 필요가 없긴하네요 원래대로 가시죠!